f3/granite.go

```diff
@@ -782,7 +791,24 @@ func (i *instance) broadcast(round uint64, step Phase, value ECChain, ticket Tic
 // The delay duration increases with each round.
 // Returns the absolute time at which the alarm will fire.
 func (i *instance) alarmAfterSynchrony() float64 {
-	timeout := i.host.Time() + i.config.Delta + (float64(i.round) * i.config.DeltaRate)
+	delta := i.config.Delta
+
+	if i.round >= 2 {
+		delta += i.config.DeltaExtra
+	}
+
+	if i.round >= 5 {
+		delta *= i.deltaRate
+		i.deltaRate *= i.config.DeltaRate
+	}
+
+	timeout := i.host.Time() + delta
+	if i.round > 0 && i.round%5 == 0 && i.phase == CONVERGE_PHASE {
```

Comment:
This reference to CONVERGE_PHASE etc is a complete surprise here. It's not mentioned in the spec at all so don't know what the intent is. I think probably there should be a parameter to this method that is passed in from the appropriate call site, and the name and documentatio of that parameter might help describe it.

The spec doesn't say anything about "mod 5". Instead it has an exponential backoff. Which you've implemented above, and seems incompatible with this fixed period.

Reply:
In #69 the pseudocode describes waiting for the next clock tick after the first round. This only occurs once at the beginning of the round. That is why here I control that by checking the the phase is CONVERGE.

What I am implementing here is the drand clock tick every 5 rounds, as the spec says. Would it make more sense in your opinion that we signal with a parameter in the call from CONVERGE that the round is starting?

Reply:
Something that is not explicitly mentioned in the spec is that it makes more sense to first send the converge message and then wait for the clock tick than otherwise, which is another reason why the above code might seem confusing.